azurelinuxagent/common/osutil/redhat.py

```diff
@@ -140,17 +139,83 @@ def get_nm_controlled(self):
 
         return True
 
-    def publish_hostname(self, hostname):
+    def get_nic_operational_and_general_states(self, ifname):
+        """
+        Checks the contents of /sys/class/net/{ifname}/operstate and the results of 'nmcli -g general.state device show {ifname}' to determine the state of the provided interface.
+        Raises an exception if the network interface state cannot be determined.
+        """
+        filepath = "/sys/class/net/{0}/operstate".format(ifname)
+        nic_general_state_cmd = ['nmcli', '-g', 'general.state', 'device', 'show', ifname]
+        if not os.path.isfile(filepath):
+            msg = "Unable to determine primary network interface {0} state, because state file does not exist: {1}".format(ifname, filepath)
+            logger.warn(msg)
+            raise Exception(msg)
+
+        try:
+            nic_operstate = fileutil.read_file(filepath).rstrip().lower()
+            nic_general_state = shellutil.run_command(nic_general_state_cmd, log_error=True).rstrip().lower()
+            if nic_operstate != "up":
+                logger.warn("The primary network interface {0} operational state is '{1}'.".format(ifname, nic_operstate))
+            else:
+                logger.info("The primary network interface {0} operational state is '{1}'.".format(ifname, nic_operstate))
+            if nic_general_state != "100 (connected)":
+                logger.warn("The primary network interface {0} general state is '{1}'.".format(ifname, nic_general_state))
+            else:
+                logger.info("The primary network interface {0} general state is '{1}'.".format(ifname, nic_general_state))
+            return nic_operstate, nic_general_state
+        except Exception as e:
+            msg = "Unexpected error while determining the primary network interface state: {0}".format(e)
+            logger.warn(msg)
+            raise Exception(msg)
+
+    def check_and_recover_nic_state(self, ifname):
+        """
+        Checks if the provided network interface is in an 'up' state. If the network interface is in a 'down' state,
+        attempt to recover the interface by restarting the Network Manager service.
+
+        Raises an exception if an attempt to bring the interface into an 'up' state fails, or if the state
+         of the network interface cannot be determined.
+        """
+        nic_operstate, nic_general_state = self.get_nic_operational_and_general_states(ifname)
+        if nic_operstate == "down" or "disconnected" in nic_general_state:
+            logger.info("Restarting the Network Manager service to recover network interface {0}".format(ifname))
+            self.restart_network_manager()
+            # Interface does not come up immediately after NetworkManager restart. Wait 1 second before checking
+            # network interface state.
+            time.sleep(1)
```

Comment:
I wonder if this should be a little longer (3? 5?). Does the value usually come up in ms or significantly less than 1 sec? it may be worth waiting more, since this is done rarely and this is happening in the env thread, which doesn't block anything significant